### ISSUE TITLE
lstsq is inaccurate

### ISSUE THREAD

**Issue by martin-marek** [google/jax#11433]
`jax.numpy.linalg.lstsq` seems to be much less accurate than `numpy.linalg.lstsq`. I was trying to implement a random perspective transform inspired by [torchvision](https://github.com/pytorch/vision/blob/0e688ce085392e748467d0858dd5d7af0f21ee9c/torchvision/transforms/functional.py#L630). One of the steps is to find a least-squares solution of `a @ x = b`.

Both in NumPy and PyTorch (using the right args), I am able to get a sufficiently-accurate answer. However, the results from JAX are unusable for me. Here is a specific example that I was trying to solve:

```python
import numpy as np
import jax.numpy as jp

# data
a = jp.array([[122, 1141, 1, 0, 0, 0, 0, 0],
              [0, 0, 0, 122, 1141, 1, -156536, -1459977],
              [686, 1209, 1, 0, 0, 0, -493286, -869838],
              [0, 0, 0, 686, 1209, 1, -877486, -1547319],
              [626, 2, 1, 0, 0, 0, -450148, -1812],
              [0, 0, 0, 626, 2, 1, 0, 0],
              [23, 339, 1, 0, 0, 0, 0, 0],
              [0, 0, 0, 23, 339, 1, 0, 0]], dtype=jp.float32)
b = jp.array([0, 1279, 719, 1279, 719, 0, 0, 0], dtype=jp.float32)

# numpy
x = np.linalg.lstsq(a, b, rcond=None)[0]
print(np.abs(a @ x - b).sum()) # 0.0005
    
# jax
x = jp.linalg.lstsq(a, b, rcond=None)[0]
print(jp.abs(a @ x - b).sum()) # 562
```

In order to get an accurate answer in JAX, I had to use the following `lstsq` solver:

```python
# jax (custom)
lstsq = lambda a, b: jp.linalg.solve(a.T @ a, a.T @ b)
x = lstsq(a, b)
print(jp.abs(a @ x - b).sum()) # 0.005
```

Is this sort of low accuracy expected or is it a bug?

I tested this in JAX 0.3.14 on CPU and 0.3.8 on GPU and TPU.

**Comment by LoicRaillon** [comment]
I think you should enable the 64-bit precision to match the Numpy precision.

```python
import jax
jax.config.update("jax_enable_x64", True)
```

**Comment by martin-marek** [comment]
Thank you, this solves the problem on CPU/GPU. Using float64, I get the following errors:
NumPy: `3.5e-09`
JAX: `4.7e-09`
Custom: `5e-12`

However, I intend to run the code on TPUs, so float64 is not really an option.

In PyTorch, for example, I can get good-enough results with float32 when I set `driver=gels`:
```python
import torch
at = torch.tensor(a.tolist(), dtype=torch.float32)
bt = torch.tensor(b.tolist(), dtype=torch.float32)
for driver in ['gelsy', 'gels', 'gelsd', 'gelss']:
    xt = torch.linalg.lstsq(at, bt, driver=driver)[0]
    error = torch.abs(at @ xt - bt).sum()
    print(driver, float(error))
# gelsy 3198
# gels 0.001
# gelsd 562
# gelss 562
```

Perhaps the results in JAX are not a bug, I was simply surprised that they're so different from NumPy. Do you think this could at least be mentioned in the [docs](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.linalg.lstsq.html)?

**Comment by LoicRaillon** [comment]
You're welcome. 

There is a section related to 64-precision in the [doc](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#double-64bit-precision).

**Comment by hawkinsp** [comment]
I suspect this comes down to the algorithm used to compute the least squares solution. Currently we use a SVD to do so:
https://github.com/google/jax/blob/b666f665ece2b91e7c07c65d27cf485a779638b6/jax/_src/numpy/linalg.py#L571

This is one of the"textbook answers on how to compute the solution to a least squares solution; the others are based on Cholesky decomposition of the `a.T @ a` matrix (much as you have done) and based on QR decomposition. The solution based on computing `a.T @ a` is usually supposed to be *less* accurate, because it "squares" the condition number of the matrix.

It seems for your matrix the SVD-based methods, both ours and those of LAPACK are inaccurate. I wonder why! Hopefully one of our numerical experts can chime in.

(Aside: On CPU and GPU, we should probably call LAPACK and Cusolver to compute least squares rather than rolling our own. The main thing missing that would enable us to do this is that we would need to define gradients for the least-squares computation directly. It would then not be difficult to make `lstsq` a primitive call to a LAPACK function: indeed the same ones that NumPy/Scipy/PyTorch use.

Since you mentioned that you are interested TPU, this isn't an option: there's no other library to resort to.)

On TPU I would be tempted to try the QR-decomposition based version which incidentally is not dissimilar to how the LAPACK function `sgels` works. (See http://www.netlib.org/lapack/explore-html/d0/db8/group__real_g_esolve_ga166c189c7bcd808a9468c05e53da816f.html for the details of the LAPACK function):

For a square matrix it looks something like this:
```
def lstsq(a, b):
  q, r = jp.linalg.qr(a)
  z = q.T @ b
  return lax.linalg.triangular_solve(r, z, left_side=True)
```

It might make sense to allow the user to select the algorithm, as Scipy and Pytorch do.

**Comment by amir-saadat** [comment]
> _both ours and those of LAPACK, are inaccurate_
This could be because we are using single precision for a computationally intense operation and you are losing accuracy quickly as a result. For example, the answer in this thread compares the FLOPS for SVD, QR, and Cholesky (https://math.stackexchange.com/questions/51150/exact-computational-costs-flop-count-for-algorithms).

**Comment by hawkinsp** [comment]
@amir-saadat I agree that double precision would be the most popular choice here. Certainly most of the time NumPy will silently upcast and compute in double precision. We intentionally do not do this in JAX because it is a very unfriendly thing to do on accelerators.

It does not however explain why we see such different precision behavior between different algorithms computed *at the same precision*.

**Comment by nvlcambier** [comment]
More data points, mostly echoing the above: this is more of an algorithm question than a CPU/GPU/TPU question imho.

`A` is very ill-conditioned. Its singular values go from 2e6 down to 4e-1
```
[2.4730732e+06 6.3774088e+05 1.7518223e+03 7.7507446e+02 5.2085254e+02
 3.9505237e+02 8.7445968e-01 4.3333033e-01]
```

The SVD of A itself is fairly accurate, even for JAX: `||A - USV^T||/||A||` is about 1e-6. This is expected: the SVD can be compute accurately even for ill-conditioned / singular matrices. 

But this does not mean that all singular values are accurate. Only the very large ones are well resolved, with near full accuracy. The very small one probably have at most 1 digit of accuracy, if not none.

JAX uses the SVD directly (pretty much the textbook algorithm, as mentioned above) for computing the minimal norm least square. But this requires inverting the singular values (larger than `rcond`). Since the smallest singular values are the least accurate, this is probably pretty inaccurate.

Numpy should be using gelsd https://www.smcm.iqfr.csic.es/docs/intel/mkl/mkl_manual/lse/functn_gelsd.htm
It's possible that gelsd has better accuracy / properties and is less sensible to the above, although I am not sure how it is implemented.

----

For the sake of experiments, consider adding this to jaxlib. 
With that, adding `CUSOLVER_INCREASE_PRECISION=1` will force cuSolver to run one more iteration to get slightly better accuracy.

```
diff --git a/jaxlib/cuda/cusolver_kernels.cc b/jaxlib/cuda/cusolver_kernels.cc
index 7ff986617..6af4e6fde 100644
--- a/jaxlib/cuda/cusolver_kernels.cc
+++ b/jaxlib/cuda/cusolver_kernels.cc
@@ -20,6 +20,8 @@ limitations under the License.
 #include <stdexcept>
 #include <utility>
 #include <vector>
+#include <cstdio>
+#include <limits>

 #include "absl/status/status.h"
 #include "absl/status/statusor.h"
@@ -859,9 +861,19 @@ static absl::Status Gesvdj_(cudaStream_t stream, void** buffers,
         float* s = static_cast<float*>(buffers[2]);
         float* u = static_cast<float*>(buffers[3]);
         float* v = static_cast<float*>(buffers[4]);
+        if(std::getenv("CUSOLVER_INCREASE_PRECISION") != nullptr) {
+          JAX_RETURN_IF_ERROR(JAX_AS_STATUS(cusolverDnXgesvdjSetTolerance(params, std::numeric_limits<float>::epsilon())));
+          JAX_RETURN_IF_ERROR(JAX_AS_STATUS(cusolverDnXgesvdjSetMaxSweeps(params, 1000)));
+        }
         JAX_RETURN_IF_ERROR(JAX_AS_STATUS(cusolverDnSgesvdj(
             handle.get(), d.jobz, d.econ, d.m, d.n, a, d.m, s, u, d.m, v,
             d.n, static_cast<float*>(work), d.lwork, info, params)));
+        int executed_sweeps = -1;
+        double residual = -1.0;
+        JAX_RETURN_IF_ERROR(JAX_AS_STATUS(cusolverDnXgesvdjGetSweeps(handle.get(), params, &executed_sweeps)));
+        JAX_RETURN_IF_ERROR(JAX_AS_STATUS(cusolverDnXgesvdjGetResidual(handle.get(), params, &residual)));
+        printf("Dgesvdj residual = %e\n", residual);
+        printf("Dgesvdj number of executed sweeps = %d \n", executed_sweeps);
         break;
       }
       case CusolverType::F64: {
```

Also consider 
```
# bug_11433.py

import numpy as np
import jax.numpy as jp

# data
a = jp.array([[122, 1141, 1, 0, 0, 0, 0, 0],
              [0, 0, 0, 122, 1141, 1, -156536, -1459977],
              [686, 1209, 1, 0, 0, 0, -493286, -869838],
              [0, 0, 0, 686, 1209, 1, -877486, -1547319],
              [626, 2, 1, 0, 0, 0, -450148, -1812],
              [0, 0, 0, 626, 2, 1, 0, 0],
              [23, 339, 1, 0, 0, 0, 0, 0],
              [0, 0, 0, 23, 339, 1, 0, 0]], dtype=jp.float32)
b = jp.array([0, 1279, 719, 1279, 719, 0, 0, 0], dtype=jp.float32)

def residual(a, b, x):
  return np.linalg.norm(np.array(a) @ np.array(x) - np.array(b)) / np.linalg.norm(np.array(b))

# numpy
x = np.linalg.lstsq(a, b, rcond=None)[0]
print("Numpy (dgelsd): ", residual(a, b, x))
    
# jax
x = jp.linalg.lstsq(a, b, rcond=None)[0]
print("JAX (sgesvdj): ", residual(a, b, x))

# jax with 0 rcond
x = jp.linalg.lstsq(a, b, rcond=0)[0]
print("JAX (sgesvdj + rcond=0): ", residual(a, b, x))
```

Then

[1] Vanilla: Numpy 8e-8, JAX 1e-1
```
$ python3 bug_11433.py
Numpy (dgelsd):  8.859738e-08
Dgesvdj residual = 6.153713e-07
Dgesvdj number of executed sweeps = 4
JAX (sgesvdj):  0.12339112 
...
JAX (sgesvdj + rcond=0):  0.26171464
```

[2] Increasing cuSOLVER's accuracy: Numpy 8e-8, JAX 1e-1 (rcond=None) or 3e-4 (rcond=0)
```
$ CUSOLVER_INCREASE_PRECISION=1 python3 bug_11433.py
Numpy (dgelsd):  8.859738e-08
Dgesvdj residual = 3.484288e-09
Dgesvdj number of executed sweeps = 5
JAX (sgesvdj):  0.12339121
...
JAX (sgesvdj + rcond=0):  0.0003707072
```

[3] Using JAX on CPU: Numpy 8e-8, JAX 1e-1 (rcond=None), JAX 4e-3 (rcond=0)
```
$ CUDA_VISIBLE_DEVICES="" python3 bug_11433.py
Numpy (dgelsd):  8.859738e-08
JAX (sgesvdj):  0.123391196
JAX (sgesvdj + rcond=0):  0.005672992
```

So
- Increasing the accuracy of the SVD
- *and* decreasing `rcond` (to capture more singular values) 
give a better solution.
